user-lock-manager-w-keypad-support.smartapp.groovy

```diff
@@ -1293,3 +1340,154 @@ def populateDiscovery(codeData, lock) {
+}
+
+def codeEntryHandler(evt) {
+	//do stuff
+    log.debug "Caught code entry event! ${evt.value.value}"
+    
+    def codeEntered = evt.value as String
+    
+    def data = evt.data as String
+    def armMode = ''
+    def currentarmMode = keypad.currentValue("armMode")
+    def changedMode = 0
+    
+    if (data == '0') {
+    	armMode = 'off'
+    }
+    else if (data == '3') {
+    	armMode = 'away'
+    }
+    else if (data == '1') {
+    	armMode = 'stay'
+    }
+    else if (data == '2') {
+    	armMode = 'stay' //Currently no separate night mode for SHM, set to 'stay'
+    }
+    else {
+    	log.error "${app.label}: Unexpected arm mode sent by keypad!: "+data
+        return []
+        }
+        
+        def i = 0
+        def message = " "
+        
+        i = settings.maxUsers
+    while (i > 0) {
+    
+    log.debug "i =" + i
+    def correctCode = settings."userCode${i}" as String
+        
+    if (codeEntered == correctCode) {
+    
+    log.debug "User Enabled: " + state."userState${i}".enabled
+    
+    if (state."userState${i}".enabled == true) {
+    	log.debug "Correct PIN entered. Change SHM state to ${armMode}"  
+        //log.debug "Delay: ${armDelay}"
+        //log.debug "Data: ${data}"
+        //log.debug "armMode: ${armMode}"
+       
+        def unlockUserName = settings."userName${i}"
+        
+        if (data == "0") {
+        	//log.debug "sendDisarmCommand"
+        	runIn(0, "sendDisarmCommand")   
+        	message = "${evt.displayName} was disarmed by ${unlockUserName}"
+        }
+        else if (data == "1") {
+        	//log.debug "sendStayCommand"
+        	runIn(armDelay, "sendStayCommand")    
+        	message = "${evt.displayName} was armed to 'Stay' by ${unlockUserName}"
+        }
+        else if (data == "2") {
+        	//log.debug "sendNightCommand"
+        	runIn(armDelay, "sendNightCommand")    
+        	message = "${evt.displayName} was armed to 'Night' by ${unlockUserName}"
+        }
+        else if (data == "3") {
+        	//log.debug "sendArmCommand"
+        	runIn(armDelay, "sendArmCommand")    
+        	message = "${evt.displayName} was armed to 'Away' by ${unlockUserName}"
+        }
+        
+        if(settings."burnCode${i}") {
+        	state."userState${i}".enabled = false
+        	message += ".  Now burning code."
+        }        
+                
+        log.debug "${message}"
+        //log.debug "Initial Usage Count:" + state."userState${i}".usage
+        state."userState${i}".usage = state."userState${i}".usage + 1
+        //log.debug "Final Usage Count:" + state."userState${i}".usage
+        send(message)
+        i = 0
+    	}
+        else if (state."userState${i}".enabled == false){
+    		log.debug "PIN Disabled"
+        	//Could also call acknowledgeArmRequest() with a parameter of 4 to report invalid code. Opportunity to simplify code?
+    		//keypad.sendInvalidKeycodeResponse()
+    	}
+    }
+    changedMode = 1
+    i--
```

Comment:
Ah, I see it now.  Our minds are just going in opposite directions.
```
for (int i = 1; i <= settings.maxUsers; i++) {
}
```
Is the same thing, but sets the variable and INCREMENTS the index through the loop.  I think this is easier to read since the logic is all in one place, but either one works.  Carry on.